File: binary_tree/988_smallest_string_starting_from_leaf.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function smallestFromLeaf(root: TreeNode | null): string {

    let ans = null;

    function traverse(node: TreeNode|null, curr: string){

        curr = String.fromCharCode(97 + node.val) + curr;

        if(!node.left && !node.right){
            if(!ans || ans > curr){
                ans = curr;
            }
        }

        if(node.left){
            traverse(node.left, curr);
        }
        if(node.right){
            traverse(node.right, curr);
        }
    }

    traverse(root, "")

    return ans

};

File: binary_tree/988_smallest_string_starting_from_leaf.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn smallest_from_leaf(root: Option<Rc<RefCell<TreeNode>>>) -> String {

        let mut ans = String::new();

        pub fn traverse(node: Option<Rc<RefCell<TreeNode>>>, mut curr:String, ans:&mut String) {

            if let Some(node) = node {

                let node_b = node.borrow();

                curr = ((node_b.val as u8 + 97) as char).to_string() + &curr;

                if node_b.left.is_none() && node_b.right.is_none() {
                    if ans.is_empty() || curr < *ans {
                        *ans = curr.clone();
                    }
                }

                if node_b.left.is_some() {
                    traverse(node_b.left.clone(), curr.clone(), ans);
                }

                if node_b.right.is_some() {
                    traverse(node_b.right.clone(), curr.clone(), ans);
                }
            }
        }

        traverse(root, String::new(), &mut ans);

        ans

    }
}